src/internal/AsyncSubject.ts

```diff
@@ -10,40 +10,28 @@ import { Subscription } from './Subscription';
  */
 export class AsyncSubject<T> extends Subject<T> {
   private value: T | null = null;
-  private hasNext: boolean = false;
-  private hasCompleted: boolean = false;
+  private hasValue = false;
 
-  /** @deprecated This is an internal implementation detail, do not use. */
-  _subscribe(subscriber: Subscriber<any>): Subscription {
-    if (this.hasError) {
-      subscriber.error(this.thrownError);
-      return Subscription.EMPTY;
-    } else if (this.hasCompleted && this.hasNext) {
-      subscriber.next(this.value);
+  protected checkFinalizedStatuses(subscriber: Subscriber<T>) {
+    const { hasError, hasValue, value, thrownError, isStopped } = this;
+    if (hasError) {
+      subscriber.error(thrownError);
+    } else if (isStopped) {
+      hasValue && subscriber.next(value!);
       subscriber.complete();
-      return Subscription.EMPTY;
     }
-    return super._subscribe(subscriber);
   }
 
   next(value: T): void {
-    if (!this.hasCompleted) {
+    if (!this.isStopped) {
       this.value = value;
-      this.hasNext = true;
-    }
-  }
-
-  error(error: any): void {
-    if (!this.hasCompleted) {
-      super.error(error);
+      this.hasValue = true;
     }
   }
 
   complete(): void {
-    this.hasCompleted = true;
-    if (this.hasNext) {
-      super.next(this.value!);
-    }
+    const { hasValue, value } = this;
+    hasValue && super.next(value!);
```

Comment:
Since `this.isStopped` (which used to be `hasCompleted`) is no longer set before doing this `super.next(value!)` is it possible this might trip something up in a reentrancy scenario since it will be incorrectly still `false` until `super.complete()` is called.

Reply:
Good catch. I've fixed that and added a test.